Add linter rules rationale (#621)
apollographql/docs: src/content/graphos/delivery/linter-rules.mdx

- Enums
- Unions

<RuleExpansionPanel
  title="TYPE_NAMES_SHOULD_BE_PASCAL_CASE"
  whatItDoes="Checks that all type names use `PascalCase`."
  rationale="`PascalCase` is a common convention for type names in GraphQL."
>

The following example violates the rule:

```graphql title="❌ schema.graphql" disableCopy=true showLineNumbers=false
# highlight-start
type streamingService { # camelCase
  # highlight-end
  id: ID!
}
```

<br/>
Use instead:
<br/>

```graphql title="✅ schema.graphql" disableCopy=true showLineNumbers=false
# highlight-start
type StreamingService { # PascalCase
  # highlight-end
  id: ID!
}
```

</RuleExpansionPanel>

<RuleExpansionPanel
  title="TYPE_PREFIX"
  whatItDoes="Checks that type names don't start with the prefix `Type`."

>   rationale="Starting type names with `Type` is redundant. It makes your schema definitions longer than necessary and your schema less readable."

Author: https://github.com/apollographql/docs/pull/621/commits/ce8eb1c345ed3737484bd3450ab7a9d2e69aebb7

>

The following example violates the rule:

```graphql title="❌ schema.graphql" disableCopy=true showLineNumbers=false
type TypeBook { # highlight-line
  title: String!
}
```

<br/>
Use instead:
<br/>

```graphql title="✅ schema.graphql" disableCopy=true showLineNumbers=false
type Book { # highlight-line
  title: String!
}
```

</RuleExpansionPanel>

<RuleExpansionPanel
  title="TYPE_SUFFIX"
  whatItDoes="Checks that type names don't use the suffix `Type`."

>   rationale="Ending type names with `Type` is redundant. It makes your schema definitions longer than necessary and your schema less readable."

Author: https://github.com/apollographql/docs/pull/621/commits/ce8eb1c345ed3737484bd3450ab7a9d2e69aebb7

These rules apply to the entire schema.

<RuleExpansionPanel
  title="DOES_NOT_PARSE"
  whatItDoes="Checks for malformed GraphQL schemas. Check your schema for any syntax errors."
/>

<RuleExpansionPanel
  title="ALL_ELEMENTS_REQUIRE_DESCRIPTION"
  whatItDoes="Checks that each element in the schema includes a [description](/apollo-server/schema/schema/#descriptions-docstrings)."
  rationale="Descriptions serve as documentation for your GraphQL schema. They make it easier for consumers of your graph discover fields and learn how to use them. Check out [examples](/apollo-server/schema/schema/#descriptions-docstrings)."
>

The following example violates the rule:

```graphql title="❌ schema.graphql" disableCopy=true showLineNumbers=false
type User {
  username: String!
}
```

<br/>
Use instead:
<br/>

```graphql title="✅ schema.graphql" disableCopy=true showLineNumbers=false
"Represents a user"
type User {

>   "The user's username"
>   username: String!

Reviewer: `username` is self-documenting. Can we use a different example or change the description:

 ```suggestion
  "A username must be [8-64] characters."
  username: String!
```

}
```

</RuleExpansionPanel>

<RuleExpansionPanel
  title="DEFINED_TYPES_ARE_USED"

>   whatItDoes="Checks that every type defined in a schema is be used at least once."
>   rationale="Unused types commonly appear after you refactor other parts of your schema. You should remove unused types to prevent confusion."

Reviewer: ```suggestion
  whatItDoes="Checks that every type defined in a schema is be used at least once."
  rationale="Unused types waste resources and should be immediately removed once they've been refactored out of a schema."
```

}
```

</RuleExpansionPanel>

<RuleExpansionPanel
  title="QUERY_DOCUMENT_DECLARATION"

>   whatItDoes="Checks that schemas don't define operations such as queries and mutations."
>   rationale="Schemas define types, fields, and their relationships. Operations specify what data to fetch or change. Since what to fetch or change is a client concern, operations should be defined on the client side. This separation allows for a clean and modular architecture, making it easier to manage and evolve the API over time."

Reviewer: ```suggestion
  whatItDoes="Checks that schemas don't define operations, such as queries and mutations."
  rationale="Operations should be defined on the client side, not within schemas. Schemas define types, fields, and their relationships. Operations specify what data to fetch or change. Since what to fetch or change is a client concern, this separation allows for a clean and modular architecture, making it easier to manage and evolve the API over time."
```
